pkg/agent/agent_windows.go

```diff
@@ -32,3 +40,99 @@ func (i *Initializer) setupExternalConnectivity() error {
 	}
 	return nil
 }
+
+// prepareHostNetworking creates HNS Network for containers.
+func (i *Initializer) prepareHostNetworking() error {
+	return util.PrepareHostNetworking(i.nodeConfig.PodCIDR, i.nodeConfig.NodeIPAddr)
+}
+
+// prepareOVSBridge adds local port and uplink to ovs bridge.
+func (i *Initializer) prepareOVSBridge() error {
+	hnsNetwork, err := hcsshim.GetHNSNetworkByName(util.LocalHNSNetwork)
+	if err != nil {
+		return err
+	}
+
+	// Set datapathID of OVS bridge.
+	// If no datapathID configured explicitly, the reconfiguration operation will change OVS bridge datapathID
+	// and break the OpenFlow channel.
+	datapathID := strings.Replace(hnsNetwork.SourceMac, ":", "", -1)
+	datapathID = "0000" + datapathID
+	if err = i.ovsBridgeClient.SetDatapathID(datapathID); err != nil {
+		klog.Errorf("Failed to set datapath_id %s: %v", datapathID, err)
+		return err
+	}
+
+	// Create local port.
+	brName := i.ovsBridgeClient.GetBridgeName()
+	if _, err := i.ovsBridgeClient.GetOFPort(brName); err == nil {
+		klog.Infof("OVS bridge local port %s already exists, skip the configuration", brName)
+	} else {
+		if _, err := i.ovsBridgeClient.CreateInternalPort(brName, config.AutoAssignedOFPort, nil); err != nil {
+			return err
+		}
+	}
+
+	// If uplink is already exists, return.
+	uplink := hnsNetwork.NetworkAdapterName
+	i.nodeConfig.UplinkName = uplink
+	if _, err := i.ovsBridgeClient.GetOFPort(uplink); err == nil {
+		klog.Infof("Uplink %s already exists, skip the configuration", uplink)
+		return err
+	}
+	// Create uplink port.
+	uplinkPortUUId, err := i.ovsBridgeClient.CreateUplinkPort(uplink, config.UplinkOFPort, nil)
+	if err != nil {
+		klog.Errorf("Failed to add uplink port %s: %v", uplink, err)
+		return err
+	}
+	uplinkInterface := interfacestore.NewUplinkInterface(uplink)
+	uplinkInterface.OVSPortConfig = &interfacestore.OVSPortConfig{uplinkPortUUId, config.UplinkOFPort}
+	i.ifaceStore.AddInterface(uplinkInterface)
+
+	// Get IP, MAC of uplink interface
+	ipAddr, ipNet, err := net.ParseCIDR(hnsNetwork.ManagementIP)
+	if err != nil {
+		klog.Errorf("Failed to parse IP Address %s for HNSNetwork %s: %v",
+			hnsNetwork.ManagementIP, util.LocalHNSNetwork, err)
+		return err
+	}
+	ifIpAddr := net.IPNet{IP: ipAddr, Mask: ipNet.Mask}
+	klog.Infof("Found hns network management ipAddr: %s", ifIpAddr.String())
+	// Move IP, MAC of from uplink interface to OVS bridge
```

Comment:
I asked this before but forgot - do we need to move the IP back to the physical interface when OVS is stopped to make sure Node network will not be down? If we need so, should we restore the IP in the Windows Service code which handles host network restoration when OVS is down? @wenyingd

Reply:
As long as the OVS bridge interface is not removed, we don't need to move the IP back to the uplink interface. As the OVS bridge has been configured with the correct address, when ovs-switchd is down, we only need to disable OVS Extension, then the interface is working like a normal interface on Windows host.

When a Transparent HNSNetwork is created, the uplink interface is updated and not able to provide the forwarding capability. The forwarding capability is returned to the uplink interface only when the HNS Network is removed. And then the IP is returned to the EthernetX automatically by Windows HNS service , so we don't need to move back in Antrea code either.

Reply:
Ok. Would you add comments for this behavior to code (not sure better in Agent code or Service code)?

Reply:
Sure, I added comments in agent codes for the description.